File: GithubGists/app/src/main/java/ru/avystavkin/githubgists/repository/transformer/TransformerHelper.java
package ru.avystavkin.githubgists.repository.transformer;

import android.text.TextUtils;

import java.util.ArrayList;
import java.util.List;

import ru.avystavkin.githubgists.models.local.Gist;
import ru.avystavkin.githubgists.models.local.GistCommit;
import ru.avystavkin.githubgists.models.local.GistCommitsHistory;
import ru.avystavkin.githubgists.models.local.GistFileInfo;
import ru.avystavkin.githubgists.models.local.User;
import ru.avystavkin.githubgists.models.server.GistCommitServer;
import ru.avystavkin.githubgists.models.server.GistCommitsHistoryServer;
import ru.avystavkin.githubgists.models.server.GistServer;
import ru.avystavkin.githubgists.models.server.UserServer;

public class TransformerHelper {

    private TransformerHelper(){}

    public static GistCommit convertGistCommitsHistory(GistCommitServer commitServer) {
        GistCommit commit = new GistCommit();
        GistCommitsHistory history = new GistCommitsHistory();

        User user = convertUser(commitServer.getUser());

        GistCommitsHistoryServer historyServer = commitServer.getStatus();
        if (historyServer != null) {
            history.setAdditions(historyServer.getAdditions());
            history.setDeletions(historyServer.getDeletions());
            history.setTotal(historyServer.getTotal());
        }
        commit.setUser(user);
        commit.setUrl(commitServer.getUrl());
        commit.setCommitDateTime(commitServer.getCommittedAt());
        commit.setCommitsHistory(history);
        commit.setVersion(commitServer.getVersion());

        return commit;
    }

    public static User convertUser(UserServer userServer) {
        User user = new User();
        if (userServer != null) {
            user.setUrl(userServer.getAvatarUrl());
            user.setId(userServer.getId());
            user.setName(userServer.getLogin());
        }
        return user;
    }

    public static Gist convertGistServer(GistServer gistServer) {
        Gist gist = new Gist();

        User user = convertUser(gistServer.getUser());

        gist.setId(gistServer.getId());

        List<GistFileInfo> list = new ArrayList<GistFileInfo>();
        for(String key : gistServer.getFiles().keySet()) {
            list.add(gistServer.getFiles().get(key));
        }
        gist.setListFiles(list);
        gist.setUrl(gistServer.getUrl());
        gist.setName(TextUtils.isEmpty(gistServer.getDescription()) ? "gist:" + gistServer.getId() : gistServer.getDescription());
        gist.setUser(user);

        return gist;
    }
}

File: GithubGists/app/src/main/java/ru/avystavkin/githubgists/repository/github/transformer/TransformerHelper.kt
package ru.avystavkin.githubgists.repository.github.transformer

import ru.avystavkin.githubgists.models.database.*
import ru.avystavkin.githubgists.models.server.GistCommitServer
import ru.avystavkin.githubgists.models.server.GistServer
import ru.avystavkin.githubgists.models.server.UserServer
import java.util.*

object TransformerHelper {

    fun convertGistCommitsHistory(commitServer: GistCommitServer): GistCommit {
        val commit = GistCommit()
        val history = GistCommitsHistory()

        val historyServer = commitServer.status
        if (historyServer != null) {
            history.additions = historyServer.additions
            history.deletions = historyServer.deletions
            history.total = historyServer.total
        }
        commit.user = convertUser(commitServer.user)
        commit.url = commitServer.url
        commit.commitDateTime = commitServer.committedAt
        commit.commitsHistory = history
        commit.version = commitServer.version

        return commit
    }

    private fun convertUser(userServer: UserServer?): User {
        val user = User()
        if (userServer != null) {
            user.url = userServer.avatarUrl
            user.id = userServer.id
            user.name = userServer.login
        }
        return user
    }

    fun convertGistServer(gistServer: GistServer): Gist {
        val gist = Gist()
        val user = convertUser(gistServer.user)

        gist.id = gistServer.id

        val list = ArrayList<GistFileInfo>()
        val size = gistServer.files?.size;
        if (size != null && size > 0) {
            for (key in gistServer.files.keys) {
                val fileInfoServer = gistServer.files[key] ?: continue
                val fileInfo = GistFileInfo()

                fileInfo.content = fileInfoServer.content
                fileInfo.fileName = fileInfoServer.fileName

                list.add(fileInfo)
            }
        }
        gist.listFiles = list
        gist.url = gistServer.url
        gist.name = if (gistServer.description.isNullOrEmpty()) "gist:" + gistServer.id!! else gistServer.description
        gist.user = user
        gist.userId = user.id

        return gist
    }
}
